src/coap_gnutls.c

```diff
@@ -736,6 +764,89 @@ static int cert_verify_callback_gnutls(gnutls_session_t g_session)
   return 0;
 }
 
+#if (GNUTLS_VERSION_NUMBER >= 0x030606)
+/* first part of Raw public key, the is the start of the Subject Public Key */
+static const unsigned char cert_asn1_header1[] = {
+  0x30, 0x59, /* SEQUENCE, length 89 bytes */
+    0x30, 0x13, /* SEQUENCE, length 19 bytes */
+      0x06, 0x07, /* OBJECT IDENTIFIER ecPublicKey (1 2 840 10045 2 1) */
+        0x2A, 0x86, 0x48, 0xCE, 0x3D, 0x02, 0x01,
+};
+/* PrimeX will get inserted */
+#if 0
+      0x06, 0x08, /* OBJECT IDENTIFIER prime256v1 (1 2 840 10045 3 1 7) */
+        0x2A, 0x86, 0x48, 0xCE, 0x3D, 0x03, 0x01, 0x07,
+#endif
+static const unsigned char cert_asn1_header2[] = {
+      0x03, 0x42, /* BIT STRING, length 66 bytes */
+/* Note: 0 bits (0x00) and no compression (0x04) are already in the certificate */
+};
+static gnutls_datum_t *
+get_asn1_tag(uint32_t tag, const uint8_t *data, size_t size)
+{
+  gnutls_datum_t *the_tag = NULL;
+  struct asn1_der_iterator i;
+  enum asn1_iterator_result result;
+  result = asn1_der_iterator_first(&i, size, data);
+  while (result != ASN1_ITERATOR_ERROR &&
+         result != ASN1_ITERATOR_END) {
+    if (i.type & ASN1_TYPE_CONSTRUCTED) {
+      the_tag = get_asn1_tag(tag, i.data, i.length);
+      if (the_tag)
+        return the_tag;
+    }
+    if ((i.type & 0xff) == tag) {
+      uint8_t *tmp = gnutls_malloc(sizeof(gnutls_datum_t) +
+                                   i.length);
+      the_tag = (gnutls_datum_t *)tmp;
+      if (the_tag == NULL)
+        return NULL;
+      the_tag->size = i.length;
+      the_tag->data = &tmp[sizeof(gnutls_datum_t)];
+      memcpy(the_tag->data, i.data, i.length);
+      return the_tag;
+    }
+    result = asn1_der_iterator_next(&i);
+  }
+  return NULL;
+}
+static gnutls_datum_t *
+get_asn1_spki(const uint8_t *data, size_t size)
+{
+  gnutls_datum_t *pub_key = get_asn1_tag(ASN1_BITSTRING, data, size);
+  gnutls_datum_t *prime = get_asn1_tag(ASN1_IDENTIFIER, data, size);
+
```

Comment:
With an additional  "get_asn1_spki_ec_private_key()" 

```
gnutls_datum_t *pub_key = get_asn1_tag(ASN1_OCTETSTRING, data, size);
...
```
 
the gnutls variant would also be able to read the ec private/public key from the openssl output.

Reply:
@boaks - It is good to have another set of eyes checking things out.  However, this is already supported and working for me with
````
            if (gnutls_pem_base64_decode2("EC PRIVATE KEY", &key,
                                          &der_private) == 0) {
              gnutls_datum_t *spki = get_asn1_spki(der_private.data,
                                                  der_private.size);

              if (spki) {
                ret = gnutls_certificate_set_rawpk_key_mem(*pki_credentials,
                                         spki,
                                         &der_private,
                                         GNUTLS_X509_FMT_DER, NULL,
                                         COAP_GNUTLS_KEY_RPK,
                                         NULL, 0, 0);
                if (ret >= 0) {
                  have_done_key = 1;
                }
                gnutls_free(spki);
              }
              gnutls_free(der_private.data);
            }
````
as `gnutls_certificate_set_rawpk_key_mem()` seems to be happy with the whole of `der_private` being passed in as the 3rd parameter.

However, `asn1_der_iterator_first()` and `asn1_der_iterator_next() `require the nettle library (which gnutls does use, but can be inbuilt), and so builds using cmake may fail - (travis build is fine) - looking into this one.  I may have to take out the hand crafted `asn1 `functions out of `src/coap_tinydtls.c` and use them instead.

Reply:
Code pushed with coap_gnutls.c updated to use local asn1 functions.